File: ios/HPTravail/HPTravail/CelestialV2/Core/CelestialAtmosphereV2.swift
import Foundation

enum CelestialWeatherTypeV2: Equatable, Sendable {
    case clear
    case partlyCloudy
    case overcast
    case fog
    case drizzle
    case rain
    case snow
    case thunderstorm
    case unknown
}

struct CelestialAtmosphereStateV2: Equatable, Sendable {
    let solarLightLevel: Double
    let twilightLevel: Double
    let nightLevel: Double
    let starsVisibility: Double
    let constellationsVisibility: Double
    let sunVisibility: Double
    let moonVisibility: Double
    let atmosphereOpacity: Double
    let cloudCoverage: Double?
    let weatherType: CelestialWeatherTypeV2
    let weatherTransmission: Double
    let ambientStarFactor: Double
    let ambientMoonFactor: Double
}

enum CelestialAtmosphereV2 {
    static func resolve(
        snapshot: CelestialSnapshotV2,
        weather: CelestialWeatherStateV2?,
        ambient: CelestialAmbientLightStateV2
    ) -> CelestialAtmosphereStateV2 {
        let sunAltitude = snapshot.sun.altitudeDegrees
        let solarLight = smoothStep(-6, 8, sunAltitude)
        let night = 1 - smoothStep(-18, -6, sunAltitude)
        let twilight = min(1, max(0, 1 - max(solarLight, night)))

        let astronomicalStarLevel = 1 - smoothStep(-14, -4, sunAltitude)
        let type = weatherType(weather?.weatherCode)
        let transmission = weatherTransmission(weather, type)
        let ambientFactor = ambientStarFactor(ambient)
        let ambientMoonFactor = ambientMoonFactor(ambient)
        let stars = min(
            1,
            max(0, astronomicalStarLevel * transmission * ambientFactor)
        )

        let sunCloudTransmission = weather.map {
            min(1, max(0.18, 1 - min(1, max(0, $0.cloudCover)) * 0.72))
        } ?? 1
        let moonCloudTransmission = weather.map {
            min(1, max(0.08, 1 - min(1, max(0, $0.cloudCover)) * 0.88))
        } ?? 1

        let sunVisibility = min(
            1,
            max(
                0,
                CelestialHorizonTransitionV2.diskOpacity(
                    altitudeDegrees: snapshot.sun.altitudeDegrees
                ) * sunCloudTransmission * phenomenonTransmission(type)
            )
        )
        let moonVisibility = min(
            1,
            max(
                0,
                CelestialHorizonTransitionV2.diskOpacity(
                    altitudeDegrees: snapshot.moon.altitudeDegrees
                ) * moonCloudTransmission * phenomenonTransmission(type) *
                    ambientMoonFactor
            )
        )

        return CelestialAtmosphereStateV2(
            solarLightLevel: solarLight,
            twilightLevel: twilight,
            nightLevel: night,
            starsVisibility: stars,
            constellationsVisibility: stars,
            sunVisibility: sunVisibility,
            moonVisibility: moonVisibility,
            atmosphereOpacity: min(1, max(0, 1 - transmission)),
            cloudCoverage: weather?.cloudCover,
            weatherType: type,
            weatherTransmission: transmission,
            ambientStarFactor: ambientFactor,
            ambientMoonFactor: ambientMoonFactor
        )
    }

    static func weatherType(_ code: Int?) -> CelestialWeatherTypeV2 {
        switch code {
        case 0: return .clear
        case 1, 2: return .partlyCloudy
        case 3: return .overcast
        case 45, 48: return .fog
        case 51, 53, 55, 56, 57: return .drizzle
        case 61, 63, 65, 66, 67, 80, 81, 82: return .rain
        case 71, 73, 75, 77, 85, 86: return .snow
        case 95, 96, 99: return .thunderstorm
        default: return .unknown
        }
    }

    private static func weatherTransmission(
        _ weather: CelestialWeatherStateV2?,
        _ type: CelestialWeatherTypeV2
    ) -> Double {
        guard let weather else { return 1 }

        let cloud = min(1, max(0.08, 1 - min(1, max(0, weather.cloudCover)) * 0.90))
        let visibility = weather.visibilityMeters.map {
            smoothStep(300, 10_000, $0)
        } ?? 1
        return min(
            1,
            max(0.03, cloud * visibility * phenomenonTransmission(type))
        )
    }

    private static func phenomenonTransmission(_ type: CelestialWeatherTypeV2) -> Double {
        switch type {
        case .fog: return 0.36
        case .drizzle: return 0.74
        case .rain: return 0.60
        case .snow: return 0.58
        case .thunderstorm: return 0.42
        case .overcast: return 0.82
        default: return 1
        }
    }

    private static func ambientStarFactor(_ ambient: CelestialAmbientLightStateV2) -> Double {
        guard ambient.quality == .valid,
              let lux = ambient.lux,
              lux.isFinite,
              lux >= 0 else {
            return 1
        }

        let factor: Double
        if lux <= 1 {
            factor = 1.08
        } else if lux <= 50 {
            factor = lerp(1.08, 1, (lux - 1) / 49)
        } else if lux <= 1_000 {
            factor = lerp(1, 0.75, (lux - 50) / 950)
        } else if lux <= 10_000 {
            factor = lerp(0.75, 0.35, (lux - 1_000) / 9_000)
        } else if lux <= 100_000 {
            factor = lerp(0.35, 0.12, (lux - 10_000) / 90_000)
        } else {
            factor = 0.12
        }
        return min(1.08, max(0.12, factor))
    }

    private static func ambientMoonFactor(_ ambient: CelestialAmbientLightStateV2) -> Double {
        guard ambient.quality == .valid,
              let lux = ambient.lux,
              lux.isFinite,
              lux >= 0 else {
            return 1
        }

        let factor: Double
        if lux <= 50 {
            factor = 1
        } else if lux <= 1_000 {
            factor = lerp(1, 0.92, (lux - 50) / 950)
        } else if lux <= 10_000 {
            factor = lerp(0.92, 0.75, (lux - 1_000) / 9_000)
        } else if lux <= 100_000 {
            factor = lerp(0.75, 0.60, (lux - 10_000) / 90_000)
        } else {
            factor = 0.60
        }
        return min(1, max(0.60, factor))
    }

    private static func smoothStep(_ edge0: Double, _ edge1: Double, _ value: Double) -> Double {
        guard value.isFinite else { return 0 }
        if edge0 == edge1 { return value < edge0 ? 0 : 1 }
        let t = min(1, max(0, (value - edge0) / (edge1 - edge0)))
        return t * t * (3 - 2 * t)
    }

    private static func lerp(_ a: Double, _ b: Double, _ t: Double) -> Double {
        a + (b - a) * min(1, max(0, t))
    }
}

File: app/src/main/java/com/amaury/pointage/v2/engine/CelestialAtmosphereV2.kt
package com.amaury.pointage.v2.engine

import com.amaury.pointage.v2.CelestialAmbientLightQualityV2
import com.amaury.pointage.v2.CelestialAmbientLightStateV2
import com.amaury.pointage.v2.CelestialWeatherContextV2
import kotlin.math.max

enum class CelestialWeatherTypeV2 {
    CLEAR,
    PARTLY_CLOUDY,
    OVERCAST,
    FOG,
    DRIZZLE,
    RAIN,
    SNOW,
    THUNDERSTORM,
    UNKNOWN
}

data class CelestialAtmosphereStateV2(
    val solarLightLevel: Double,
    val twilightLevel: Double,
    val nightLevel: Double,
    val starsVisibility: Double,
    val constellationsVisibility: Double,
    val sunVisibility: Double,
    val moonVisibility: Double,
    val atmosphereOpacity: Double,
    val cloudCoverage: Double?,
    val weatherType: CelestialWeatherTypeV2,
    val weatherTransmission: Double,
    val ambientStarFactor: Double,
    val ambientMoonFactor: Double
)

/**
 * Propriétaire pur des coefficients atmosphériques de Céleste.
 *
 * L'astronomie reste la vérité de niveau 1. Météo et luminosité ne déplacent
 * jamais un astre : elles ne font qu'atténuer ou renforcer le rendu visuel.
 */
object CelestialAtmosphereV2 {
    fun resolve(
        snapshot: CelestialSnapshotV2,
        weather: CelestialWeatherContextV2.State?,
        ambient: CelestialAmbientLightStateV2
    ): CelestialAtmosphereStateV2 {
        val sunAltitude = snapshot.sun.altitudeDeg
        val solarLight = smoothStep(-6.0, 8.0, sunAltitude)
        val night = 1.0 - smoothStep(-18.0, -6.0, sunAltitude)
        val twilight = (1.0 - max(solarLight, night)).coerceIn(0.0, 1.0)

        val astronomicalStarLevel = 1.0 - smoothStep(-14.0, -4.0, sunAltitude)
        val weatherType = weatherType(weather?.weatherCode)
        val weatherTransmission = weatherTransmission(weather, weatherType)
        val ambientFactor = ambientStarFactor(ambient)
        val ambientMoonFactor = ambientMoonFactor(ambient)

        val stars = (
            astronomicalStarLevel * weatherTransmission * ambientFactor
            ).coerceIn(0.0, 1.0)

        val sunCloudTransmission = weather?.cloudCover
            ?.let { (1.0 - it.coerceIn(0.0, 1.0) * 0.72).coerceIn(0.18, 1.0) }
            ?: 1.0
        val moonCloudTransmission = weather?.cloudCover
            ?.let { (1.0 - it.coerceIn(0.0, 1.0) * 0.88).coerceIn(0.08, 1.0) }
            ?: 1.0

        val sunVisibility = (
            CelestialHorizonTransitionV2.diskAlpha(snapshot.sun.altitudeDeg) *
                sunCloudTransmission * phenomenonTransmission(weatherType)
            ).coerceIn(0.0, 1.0)
        val moonVisibility = (
            CelestialHorizonTransitionV2.diskAlpha(snapshot.moon.altitudeDeg) *
                moonCloudTransmission * phenomenonTransmission(weatherType) *
                ambientMoonFactor
            ).coerceIn(0.0, 1.0)

        return CelestialAtmosphereStateV2(
            solarLightLevel = solarLight,
            twilightLevel = twilight,
            nightLevel = night,
            starsVisibility = stars,
            constellationsVisibility = stars,
            sunVisibility = sunVisibility,
            moonVisibility = moonVisibility,
            atmosphereOpacity = (1.0 - weatherTransmission).coerceIn(0.0, 1.0),
            cloudCoverage = weather?.cloudCover,
            weatherType = weatherType,
            weatherTransmission = weatherTransmission,
            ambientStarFactor = ambientFactor,
            ambientMoonFactor = ambientMoonFactor
        )
    }

    fun weatherType(code: Int?): CelestialWeatherTypeV2 = when (code) {
        0 -> CelestialWeatherTypeV2.CLEAR
        1, 2 -> CelestialWeatherTypeV2.PARTLY_CLOUDY
        3 -> CelestialWeatherTypeV2.OVERCAST
        45, 48 -> CelestialWeatherTypeV2.FOG
        51, 53, 55, 56, 57 -> CelestialWeatherTypeV2.DRIZZLE
        61, 63, 65, 66, 67, 80, 81, 82 -> CelestialWeatherTypeV2.RAIN
        71, 73, 75, 77, 85, 86 -> CelestialWeatherTypeV2.SNOW
        95, 96, 99 -> CelestialWeatherTypeV2.THUNDERSTORM
        else -> CelestialWeatherTypeV2.UNKNOWN
    }

    private fun weatherTransmission(
        weather: CelestialWeatherContextV2.State?,
        type: CelestialWeatherTypeV2
    ): Double {
        if (weather == null) return 1.0

        val cloud = (1.0 - weather.cloudCover.coerceIn(0.0, 1.0) * 0.90)
            .coerceIn(0.08, 1.0)
        val visibility = weather.visibilityMeters?.let {
            smoothStep(300.0, 10_000.0, it)
        } ?: 1.0
        val phenomenon = phenomenonTransmission(type)

        return (cloud * visibility * phenomenon).coerceIn(0.03, 1.0)
    }

    private fun phenomenonTransmission(type: CelestialWeatherTypeV2): Double = when (type) {
        CelestialWeatherTypeV2.FOG -> 0.36
        CelestialWeatherTypeV2.DRIZZLE -> 0.74
        CelestialWeatherTypeV2.RAIN -> 0.60
        CelestialWeatherTypeV2.SNOW -> 0.58
        CelestialWeatherTypeV2.THUNDERSTORM -> 0.42
        CelestialWeatherTypeV2.OVERCAST -> 0.82
        else -> 1.0
    }

    private fun ambientStarFactor(ambient: CelestialAmbientLightStateV2): Double {
        if (ambient.quality != CelestialAmbientLightQualityV2.VALID) return 1.0
        val lux = ambient.lux ?: return 1.0
        if (!lux.isFinite() || lux < 0.0) return 1.0

        return when {
            lux <= 1.0 -> 1.08
            lux <= 50.0 -> lerp(1.08, 1.0, (lux - 1.0) / 49.0)
            lux <= 1_000.0 -> lerp(1.0, 0.75, (lux - 50.0) / 950.0)
            lux <= 10_000.0 -> lerp(0.75, 0.35, (lux - 1_000.0) / 9_000.0)
            lux <= 100_000.0 -> lerp(0.35, 0.12, (lux - 10_000.0) / 90_000.0)
            else -> 0.12
        }.coerceIn(0.12, 1.08)
    }

    private fun ambientMoonFactor(ambient: CelestialAmbientLightStateV2): Double {
        if (ambient.quality != CelestialAmbientLightQualityV2.VALID) return 1.0
        val lux = ambient.lux ?: return 1.0
        if (!lux.isFinite() || lux < 0.0) return 1.0

        return when {
            lux <= 50.0 -> 1.0
            lux <= 1_000.0 -> lerp(1.0, 0.92, (lux - 50.0) / 950.0)
            lux <= 10_000.0 -> lerp(0.92, 0.75, (lux - 1_000.0) / 9_000.0)
            lux <= 100_000.0 -> lerp(0.75, 0.60, (lux - 10_000.0) / 90_000.0)
            else -> 0.60
        }.coerceIn(0.60, 1.0)
    }

    private fun smoothStep(edge0: Double, edge1: Double, value: Double): Double {
        if (!value.isFinite()) return 0.0
        if (edge0 == edge1) return if (value < edge0) 0.0 else 1.0
        val t = ((value - edge0) / (edge1 - edge0)).coerceIn(0.0, 1.0)
        return t * t * (3.0 - 2.0 * t)
    }

    private fun lerp(a: Double, b: Double, t: Double): Double =
        a + (b - a) * t.coerceIn(0.0, 1.0)
}
